src/mlpack/core/tree/rectangle_tree/discrete_hilbert_value.hpp

```diff
@@ -0,0 +1,200 @@
+/**
+ * @file discrete_hilbert_value.hpp
+ * @author Mikhail Lozhnikov
+ *
+ * Defintion of the DiscreteHilbertValue class, a class that calculates
+ * the ordering of points using the Hilbert curve.
+ */
+#ifndef MLPACK_CORE_TREE_RECTANGLE_TREE_DISCRETE_HILBERT_VALUE_HPP
+#define MLPACK_CORE_TREE_RECTANGLE_TREE_DISCRETE_HILBERT_VALUE_HPP
+
+#include <mlpack/core.hpp>
+
+namespace mlpack {
+namespace tree /** Trees and tree-building procedures. */ {
+
+class DiscreteHilbertValue
+{
+ public:
+  //! Default constructor
+  DiscreteHilbertValue();
+
+  /**
+   * Construct this for the node tree. If the node is the root this method
+   * computes the Hilbert value for each point in the tree's dataset.
+   * @param node The node that stores this Hilbert value.
+   */
+  template<typename TreeType>
+  DiscreteHilbertValue(const TreeType *tree);
+
+  /**
+   * Create a Hilbert value object by copying from the other node.
+   * @param other The node from which the value will be copied.
+   */
+  template<typename TreeType>
+  DiscreteHilbertValue(const TreeType &other);
```

Comment:
This looks like a copy constructor, should it take a `const DiscreteHilbertValue&` instead of a `TreeType&`?